### ISSUE TITLE
Diverter - coils - activation_time

### ISSUE THREAD

**Issue by Maverick78440** [missionpinball/mpf#366]
version : 0.21.4 (updated from github)

I have some difficulties about activating the 2 coils of 2 diverters.

The pinball machine is the Attack from Mars. The first coil is to move the physical diverter and the second one is there to block it in the position. The first is disable after 1 sec. The second one could stay open during the complete game.

Theirs jobs are to open the way of the lock. When a ball is drained, the diverter has to be re-open and it is there where I have a problem, when I resend the event via the code to reopen it, the coil which shoult stay 1s, continue to stay activated.

Code is here.

balldevices section in the main config.yaml

```
    lock_diverter:
        activation_coil: c_diverter_power
        type: hold
        activation_time: 1s
        targets_when_inactive: playfield
        activation_switches: s_right_ramp_exit
        enable_events: mode_base_started, multiball_locks_multiball_ended
        disable_events: ball_ending, multiball_locks_multiball_started#, disable_div_power
#        activate_events: re_launch_diverter
#        deactivate_events: disable_div_power
    lock_diverter2:
        activation_coil: c_diverter_hold
        type: hold
#        activation_time: 30s
        targets_when_inactive: playfield
        activation_switches: s_center_ramp_enter
        enable_events: mode_base_started, multiball_locks_multiball_ended
        disable_events: ball_ending, multiball_locks_multiball_started
        activate_events: shot_center_ramp
#        deactivate_events: stop_lock_hold


`
coils section of the main config.yaml

`    c_diverter_power:
        number: C35
        hold_power: 5
        pulse_ms: 30
#        pulse_events: diverter_power_pulse
        enable_events: s_right_ramp_exit_hit, enable_div_power
        disable_events: multiball_locks_multiball_started, ball ending
    c_diverter_hold:
        number: C36
        pulse_ms: 30
        hold_power: 3
        allow_enable: true
        enable_events: s_center_ramp_enter_active, enable_div_hold
        disable_events: multiball_locks_multiball_started, ball_ending`

code Inside lock.py

`from mpf.system.mode import Mode

class lock(Mode):
    def mode_init(self):
        self.display = 0

        ######################################################################################
        self.machine.events.add_handler('s_left_popper_active', self.nb_lock)
        self.machine.events.add_handler('locks_multiball_start', self.locks_init)
        ######################################################################################

    def mode_start(self, **kwargs):
        if self.player['slpa'] == 0:
            self.machine.events.post('lock_0')
#            self.machine.events.post('enable_div_power')
#            self.machine.events.post('enable_div_hold')
#            self.machine.events.post('disable_div_power')

        if self.player['slpa'] == 1:
            self.machine.events.post('lock_1')
#            self.machine.events.post('s_center_ramp_enter_active')
#            self.machine.events.post('s_center_ramp_enter_inactive')            
#            self.machine.events.post('s_right_ramp_exit_active')

        if self.player['slpa'] == 2:
            self.machine.events.post('lock_2')
#            self.machine.events.post('s_center_ramp_enter_hit')
#            self.machine.events.post('s_right_ramp_exit_hit')

#        if self.player['slpa'] == 3:
#            self.machine.events.post('lock_3')

    def clear(self, **kwargs):
        self.display = 0

    def nb_lock(self, **kwargs):
        self.display = 1

    def locks_init(self, **kwargs):
        self.display = 1

        self.machine.game.player.slpa = 0

`

```

**Comment by jabdoa2** [comment]
Ok let me recap what you want to achieve: 1. You have a diverter with dual wound coil and want to activate it when the base mode started. 2. You want to deactivate it on ball end and active it again on the next ball.

There are a few issues here:
1. activation_time: 0 may to be broken in 0.21. I verified that it is fixed and tested in 0.30.
2. 0.21 does not have dual wound coil support for diverters. Generic dual wound coil support was added for 0.30. I added #368 to test it with diverters (but i expect it to work).
3. Your restart does not work because you only disable the diverter but you do not deactivate it. Therefore, it will stay activated until the activation_time expired.

To fix your problems, you can try the following:
1. Remove activation_events from your coils (or remove the diverter).
2. Add a deactivation_events line
3. Then post re_launch_diverter in you mode code to enable and activate the diverter again.

```
    lock_diverter:
        activation_coil: c_diverter_power
        type: hold
        activation_time: 1s
        activation_switches: s_right_ramp_exit
        enable_events: mode_base_started, multiball_locks_multiball_ended,re_launch_diverter
        disable_events: ball_ending, multiball_locks_multiball_started
        activate_events: re_launch_diverter|10ms
        deactivate_events: ball_ending, multiball_locks_multiball_started
    lock_diverter2:
        activation_coil: c_diverter_hold
        type: hold
        activation_switches: s_right_ramp_exit
        enable_events: mode_base_started, multiball_locks_multiball_ended,re_launch_diverter
        disable_events: ball_ending, multiball_locks_multiball_started
        activate_events: re_launch_diverter|10ms
        deactivate_events: ball_ending, multiball_locks_multiball_started

    c_diverter_power:
        number: C35
        hold_power: 5
        pulse_ms: 30
    c_diverter_hold:
        number: C36
        pulse_ms: 30
        hold_power: 3
        allow_enable: true
```

What you did was changing the coils outside the diverter. Thereby diverter state and coils no longer match. Also please make sure you understand the difference between enabling and activating a diverter. Enabled -> activates when activation_switch is hit or activation_event is posted.

**Comment by Maverick78440** [comment]
Hi,
I made all the modifications.
The 10ms seems to be not understood by mpf.

```
INFO : Machine : Mission Pinball Framework v0.21.6
INFO : Machine : Machine config file #1: c:\pinball\actarus\config\config
INFO : Machine : Loading system modules...
INFO : DeviceManager : Loading devices...
ERROR : root : invalid literal for int() with base 10: '10ms'
Traceback (most recent call last):
  File "mpf.py", line 110, in main
    machine = MachineController(vars(args))
  File "C:\pinball\mpf\mpf\system\machine.py", line 127, in __init__
    self.events._process_event_queue()
  File "C:\pinball\mpf\mpf\system\events.py", line 546, in _process_event_queue
    **event[3])
  File "C:\pinball\mpf\mpf\system\events.py", line 486, in _process_event
    result = handler[0](**merged_kwargs)
  File "C:\pinball\mpf\mpf\system\device_manager.py", line 121, in create_machinewide_device_control_events
    priority=int(priority),
ValueError: invalid literal for int() with base 10: '10ms'

C:\pinball\mpf>
```
